va/validation-authority.go

```diff
@@ -437,36 +444,53 @@ func newCAASet(CAAs []*dns.CAA) *CAASet {
 }
 
 func (va *ValidationAuthorityImpl) getCAASet(hostname string) (*CAASet, error) {
-	hostname = strings.TrimRight(hostname, ".")
-	splitDomain := strings.Split(hostname, ".")
-	// RFC 6844 CAA set query sequence, 'x.y.z.com' => ['x.y.z.com', 'y.z.com', 'z.com']
-	for i := range splitDomain {
-		queryDomain := strings.Join(splitDomain[i:], ".")
-		// Don't query a public suffix
-		if _, present := policy.PublicSuffixList[queryDomain]; present {
+	label := strings.TrimRight(hostname, ".")
+	cnames := 0
+	// See RFC 6844 "Certification Authority Processing" for pseudocode.
+	for {
+		if strings.IndexRune(label, '.') == -1 {
+			// Reached TLD
 			break
 		}
-
-		// Query CAA records for domain and its alias if it has a CNAME
-		for _, alias := range []bool{false, true} {
-			if alias {
-				target, _, err := va.DNSResolver.LookupCNAME(queryDomain)
-				if err != nil {
-					return nil, err
-				}
-				queryDomain = target
-			}
-			CAAs, _, err := va.DNSResolver.LookupCAA(queryDomain)
-			if err != nil {
-				return nil, err
-			}
-
-			if len(CAAs) > 0 {
-				return newCAASet(CAAs), nil
+		if _, present := policy.PublicSuffixList[label]; present {
+			break
+		}
+		CAAs, _, err := va.DNSResolver.LookupCAA(label)
+		if err != nil {
+			return nil, err
+		}
+		cname, _, err := va.DNSResolver.LookupCNAME(label)
```

Comment:
FWIW, I've learned that
* None of the usual dns tools are willing to look up type CAA or 257 for me at all, because they've never heard of it.
* If I publish both CNAME and A on a name, and ask 8.8.8.8 for an A record for that name, it follows the CNAME and gives me the CNAME's A record. E.g.,

```
$ host -t a a-and-cname.d.zyx.ca ns1
a-and-cname is an alias for fubar
a-and-cname has address 10.10.10.10

$ host -t a a-and-cname.d.zyx.ca 8.8.8.8
a-and-cname is an alias for fubar
fubar has address 10.0.0.2
```

If caches handle CAA+CNAME the same way, this error-checking is useless and we just have to assume/hope the RFC meant to say "follow CNAME as usual, behavior is undefined as usual if both exist".